### PR TITLE
system-helper: Use new ostree APIs to reject world-writable/suid content

### PR DESCRIPTION
**Pull request by cgwalters**
This uses the new libostree APIs that landed recently to ensure
that we reject any files with mode outside of `0775` for system
helper pulls, and we also neuter directories during checkout.

However, this does *not* fix up any already downloaded content.
For that, one could uninstall/reinstall; or a future patch could
do a one-time fixup pass.

Closes: https://github.com/flatpak/flatpak/issues/845

### CODE REVIEW

**Comment by cgwalters** [comment]
Side note; I found it double plus confusing that the system helper path actually wasn't using `flatpak_dir_pull_untrusted_local()`...why does the code try to distinguish between "local repo" and `file:///` URI?  In libostree they should be the same thing.

**Comment by cgwalters** [comment]
Also something I'm not entirely sure of is the appstream stuff; we appear to be doing pulls for that too that go through separate paths?

**Comment by alexlarsson** [comment]
It *is* using flatpak_dir_pull_untrusted_local(). There are two codepaths here. 

One is the case where the system repo has a remote configured with a file: uri. In this case we just do the pull entierly in the system helper. This is the codepath your current patch has implemented.

The other case is where the remote has a non-local uri, but the app already downloaded this to a separate local directory, which is specified with arg_repo_path. In this case we're using flatpak_dir_pull_untrusted_local(). The current patch doesn't handle this case.

These are not the same, because in one case the sysadmin defined a pathname as "trusted" by configuring a remote at that location. In the second case the caller (who is untrusted) passes in the directory in which to read the files from, and we have to be much more careful about what we pull from this. 

For the appstream we're pulling in the same way as for app, but the deploy is different (because the checkout is to a different location). This looks fine in the patch.

**Comment by cgwalters** [comment]
Ah, and because I was testing this with a "local" remote from a locally built app I ended up in the other path.  OK.

I actually did patch `flatpak_dir_pull_untrusted_local()` though.  That's what I tried first and was really confused when it didn't work.

**Comment by cgwalters** [comment]
Or in other words, the patch is correct?

**Comment by alexlarsson** [comment]
Oh, lemme check.
It needs a configure check for the new ostree version though.

**Comment by alexlarsson** [comment]
Ah, yes. It is correct. However, i would like you to also pass that flag in repo_pull_one_dir, so we get nice warnings about this earlier in the process (and for --user installs).

**Comment by cgwalters** [comment]
Let's land https://github.com/flatpak/flatpak/pull/849 first.

**Comment by cgwalters** [comment]
Like that :arrow_up:  ?

**Comment by cgwalters** [comment]
> error: Can't use bareuseronly-files with non-local origin repo

Ah, yes.  Mrm.  I had originally designed this only for the system helper but perhaps it's better to support it for HTTP pulls too.  Will make a change in libostree.

**Comment by cgwalters** [comment]
Now depends on https://github.com/ostreedev/ostree/pull/930

**Comment by alexlarsson** [comment]
I guess you can rebase this on master now.

**Comment by cgwalters** [comment]
Done. :surfing_man:

File: system-helper/flatpak-system-helper.c [hunk 1]
```diff
@@ -298,6 +298,8 @@ handle_deploy (FlatpakSystemHelper   *object,
     }
   else if (local_pull)
     {
+      const OstreeRepoPullFlags pull_flags = OSTREE_REPO_PULL_FLAGS_UNTRUSTED |
+                                             OSTREE_REPO_PULL_FLAGS_BAREUSERONLY_FILES;
```
Comment (alexlarsson) [inline system-helper/flatpak-system-helper.c]: I don't think this one is needed. The call to flatpak_dir_pull() will eventually call repo_pull_one_dir, which sets this.

Comment (cgwalters) [inline system-helper/flatpak-system-helper.c]: OK, removed in :arrow_down:

**Comment by cgwalters** [comment]
OK, let's give things a day or two at least in git master and do releases?

**Pull request merged**